File: 3.3P/Clock/Clock/Clock.cs
using System;

namespace Clock
{
	public class Clock
	{
		private Counter _hour;
		private Counter _minute;
		private Counter _second;

		public Clock()
		{
			_hour = new Counter("Hour");
			_minute = new Counter("Minute");
			_second = new Counter("Second");
		}

		public int Hour => _hour.Ticks % 24;
		public int Minute => _minute.Ticks % 60;
		public int Second => _second.Ticks % 60;

		public void Tick()
		{
			_second.Increment();
			if (Second == 0)
			{
				_minute.Increment();
				if (Minute == 0)
					_hour.Increment();
			}
		}

		public void Reset()
		{
			_hour.Reset();
			_minute.Reset();
			_second.Reset();
		}

		public string Time => $"{Hour:00}:{Minute:00}:{Second:00}";
	}
}

File: 11.1P/P11_1/src/clock.rs
use crate::counter::Counter;

pub struct Clock {
	hour: Counter,
	minute: Counter,
	second: Counter,
}

impl Clock {
	pub fn new() -> Self {
		Self {
			hour: Counter::new("Hour"),
			minute: Counter::new("Minute"),
			second: Counter::new("Second"),
		}
	}

	pub fn tick(&mut self) {
		self.second.inc();
		if self.get_second() == 0 {
			self.minute.inc();
			if self.get_minute() == 0 {
				self.hour.inc();
			}
		}
	}

	pub fn reset(&mut self) {
		self.hour.reset();
		self.minute.reset();
		self.second.reset();
	}

	pub fn get_hour(&self) -> i32 { self.hour.get_count() % 24 }
	pub fn get_minute(&self) -> i32 { self.minute.get_count() % 60 }
	pub fn get_second(&self) -> i32 { self.second.get_count() % 60 }
	pub fn get_ticks(&self) -> i32 { self.second.get_count() }

	pub fn time(&self) -> String {
		format!(
			"{:02}:{:02}:{:02}",
			self.get_hour(),
			self.get_minute(),
			self.get_second(),
		)
	}
}
